code/datums/components/attachment_handler.dm

```diff
@@ -32,254 +32,285 @@
 	attachable_overlays += slots
 
 	var/obj/parent_object = parent
-	if(length(starting_attachments) && parent_object.loc) //Attaches starting attachments if the object is not instantiated in nullspace. If it is created in null space, such as in a loadout vendor. It wont create default attachments.
+	if(parent_object.loc) //Attaches starting attachments if the object is not instantiated in nullspace. If it is created in null space, such as in a loadout vendor. It wont create default attachments.
 		for(var/starting_attachment_type in starting_attachments)
-			attach_without_user(attachment = new starting_attachment_type(parent_object))
+			attach_on_spawn(attachment = new starting_attachment_type(parent_object))
 
 	update_parent_overlay()
 
-	RegisterSignal(parent, COMSIG_PARENT_ATTACKBY, .proc/start_handle_attachment) //For attaching.
-	RegisterSignal(parent, list(COMSIG_LOADOUT_VENDOR_VENDED_GUN_ATTACHMENT, COMSIG_LOADOUT_VENDOR_VENDED_ATTACHMENT_GUN, COMSIG_LOADOUT_VENDOR_VENDED_ARMOR_ATTACHMENT), .proc/attach_without_user)
+	RegisterSignal(parent, COMSIG_PARENT_ATTACKBY, .proc/start_mob_attempt_attach) //For attaching.
+	RegisterSignal(parent, list(COMSIG_LOADOUT_VENDOR_VENDED_GUN_ATTACHMENT, COMSIG_LOADOUT_VENDOR_VENDED_ATTACHMENT_GUN, COMSIG_LOADOUT_VENDOR_VENDED_ARMOR_ATTACHMENT), .proc/attach_on_spawn)
+	RegisterSignal(parent, COMSIG_MARINE_VENDOR_MODULE_VENDED, .proc/vendor_attempt_attach)
 
-	RegisterSignal(parent, COMSIG_CLICK_ALT, .proc/start_detach) //For Detaching
+	RegisterSignal(parent, COMSIG_CLICK_ALT, .proc/start_mob_attempt_detach) //For Detaching
 	RegisterSignal(parent, COMSIG_PARENT_QDELETING, .proc/clean_references) //Dels attachments.
 	RegisterSignal(parent, COMSIG_ITEM_APPLY_CUSTOM_OVERLAY, .proc/apply_custom)
 	RegisterSignal(parent, COMSIG_ITEM_UNEQUIPPED, .proc/remove_overlay)
 
-///Starts processing the attack, and whether or not the attachable can attack.
-/datum/component/attachment_handler/proc/start_handle_attachment(datum/source, obj/attacking, mob/attacker)
-	SIGNAL_HANDLER
-	INVOKE_ASYNC(src, .proc/handle_attachment, attacking, attacker)
+///Gets data from the /datum/element/attachment on the specified item. Returns null if target isn't really an attachment.
+/datum/component/attachment_handler/proc/get_attachment_data(obj/attachment)
+	var/list/attachment_data = list()
+	SEND_SIGNAL(attachment, COMSIG_ITEM_IS_ATTACHING, attachment_data) //attachment element adds data to list if present on item
```

Comment:
Signal have a rule : they need to be sent at only one place in code.

Reply:
Resolved by me removing the attachment changes.